test/jdk/java/net/httpclient/FileChannelPublisherTest.java

```diff
@@ -0,0 +1,693 @@
+                    Thread.currentThread().interrupt();     // Restore the `interrupted` flag
+                    throw new RuntimeException(ie);
+                }
+            }
+        }
+
+        @Override
+        public String toString() {
+            return serverName;
+        }
+
+    }
+
+    @AfterAll
+    static void shutDown() {
+        LOGGER.log("Closing the client");
+        CLIENT.close();
+        LOGGER.log("Closing servers");
+        closeServers();
+        LOGGER.log("Closing the executor");
+        EXECUTOR.shutdownNow();
+    }
+
+    private static void closeServers() {
+        Exception[] exceptionRef = {null};
+        Stream
+                .of(HTTP1, HTTPS1, HTTP2, HTTPS2)
+                .map(pair -> (Runnable) pair.server::stop)
+                .forEach(terminator -> {
+                    try {
+                        terminator.run();
+                    } catch (Exception exception) {
+                        if (exceptionRef[0] == null) {
+                            exceptionRef[0] = exception;
+                        } else {
+                            exceptionRef[0].addSuppressed(exception);
+                        }
+                    }
+                });
+        if (exceptionRef[0] != null) {
+            throw new RuntimeException("failed closing one or more server resources", exceptionRef[0]);
+        }
+    }
+
+    /**
+     * Resets {@link ServerRequestPair#serverReadRequestBodyBytes()} to avoid leftover state from a test leaking to the next.
+     */
+    @BeforeEach
+    void resetServerHandlerResults() {
+        Stream
+                .of(HTTP1, HTTPS1, HTTP2, HTTPS2)
+                .forEach(pair -> pair.serverReadRequestBodyBytes.clear());
+    }
+
+    static ServerRequestPair[] serverRequestPairs() {
+        return new ServerRequestPair[]{
+                HTTP1,
+                HTTPS1,
+                HTTP2,
+                HTTPS2
+        };
+    }
+
+    @Test
+    void testNullFileChannel() {
+        assertThrows(NullPointerException.class, () -> ofFileChannel(null, 0, 1));
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "6,-1,1",   // offset < 0
+            "6,7,1",    // offset > fileSize
+            "6,0,-1",   // length < 0
+            "6,0,7",    // length > fileSize
+            "6,2,5"     // (offset + length) > fileSize
+    })
+    void testIllegalOffset(
+            int fileLength,
+            int fileChannelOffset,
+            int fileChannelLength,
+            @TempDir(cleanup = CleanupMode.ON_SUCCESS) Path tempDir) throws Exception {
+        withFileChannel(tempDir.resolve("data.txt"), fileLength, (_, fileChannel) ->
+                assertThrows(
+                        IndexOutOfBoundsException.class,
+                        () -> ofFileChannel(fileChannel, fileChannelOffset, fileChannelLength)));
+    }
+
+    @ParameterizedTest
+    @MethodSource("serverRequestPairs")
+    void testContentLessThanBufferSize(
```

Comment:
Hello Volkan, on first read, it's not clear to me what this test is testing. Can a brief comment be added to this test method noting what it is testing?

Reply:
Good call. Improved the documentation of both `testContent{Less,More}ThanBufferSize()` methods in 99503c0bac7.